feat: add "Close library" to File menu

SatyamKumarCS opened the pull request:
Closes #14381 

This PR adds three new library-management actions to the File menu:
- **Close library** – closes the currently active library.
- **Close others** – closes all libraries except the currently active one.
- **Close all** – closes all open libraries.

Previously, these actions were only available in the library tab’s context menu,
making them harder to discover. Adding them to the File menu improves usability,
aligns with common desktop application UX patterns, and provides faster access
to library management operations.

### Steps to test

1. Start JabRef.
2. Open multiple libraries (e.g., Library A, Library B, Library C).
3. Go to **File → Close library**:
   - Only the active library should close.
4. Go to **File → Close others**:
   - The active library remains open.
   - All other libraries close.
5. Go to **File → Close all**:
   - All open libraries close.
   - JabRef returns to the empty state with no libraries open.
 
<img width="1512" height="853" alt="Screenshot 2025-11-22 at 3 11 46 AM" src="https://github.com/user-attachments/assets/0423cf8d-2d4b-420c-b215-ce034c11f248" />

### Mandatory checks

<!--
Go through the checklist below. It is mandatory, even for a draft pull request.

Keep ALL the items. Replace the dots inside [.] and mark them as follows: 
[x] done 
[ ] TODO (yet to be done)
[/] not applicable
-->

- [x] I own the copyright of the code submitted and I license it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] I manually tested my changes in running JabRef (always required)
- [/] I added JUnit tests for changes (if applicable)
- [x] I added screenshots in the PR description (if change is visible to the user)
- [/] I described the change in `CHANGELOG.md` in a way that is understandable for the average user (if change is visible to the user)
- [x] I checked the [user documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request updating file(s) in <https://github.com/JabRef/user-documentation/tree/main/en>.

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -144,64 +134,23 @@ public JabRefFrame(Stage mainStage,
-                mainStage::getScene,
-                this::getOpenDatabaseAction,
-                stateManager);
+        FrameDndHandler frameDndHandler = new FrameDndHandler(tabbedPane, mainStage::getScene, this::getOpenDatabaseAction, stateManager);
```
Please keep linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -144,64 +134,23 @@ public JabRefFrame(Stage mainStage,
-                undoManager,
-                dialogService,
-                SearchType.NORMAL_SEARCH);
+        this.globalSearchBar = new GlobalSearchBar(this, stateManager, this.preferences, undoManager, dialogService, SearchType.NORMAL_SEARCH);
```
Please keep linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -144,64 +134,23 @@ public JabRefFrame(Stage mainStage,
                 // Actions are recreated here since this avoids passing more parameters and the amount of additional memory consumption is neglegtable.
-                new UndoAction(this::getCurrentLibraryTab, undoManager, dialogService, stateManager),
-                new RedoAction(this::getCurrentLibraryTab, undoManager, dialogService, stateManager));
+                new UndoAction(this::getCurrentLibraryTab, undoManager, dialogService, stateManager), new RedoAction(this::getCurrentLibraryTab, undoManager, dialogService, stateManager));
```
Please keep linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -144,64 +134,23 @@ public JabRefFrame(Stage mainStage,
-                this.preferences.getLastFilesOpenedPreferences().getFileHistory(),
-                dialogService,
-                getOpenDatabaseAction());
+        this.sidePane = new SidePane(this, this.preferences, Injector.instantiateModelOrService(JournalAbbreviationRepository.class), taskExecutor, dialogService, aiService, stateManager, entryEditor, fileUpdateMonitor, entryTypesManager, clipBoardManager, undoManager);
```
Please keep linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -144,64 +134,23 @@ public JabRefFrame(Stage mainStage,
-                getOpenDatabaseAction());
+        this.sidePane = new SidePane(this, this.preferences, Injector.instantiateModelOrService(JournalAbbreviationRepository.class), taskExecutor, dialogService, aiService, stateManager, entryEditor, fileUpdateMonitor, entryTypesManager, clipBoardManager, undoManager);
+
+        this.pushToApplicationCommand = new GuiPushToApplicationCommand(stateManager, dialogService, this.preferences, taskExecutor);
```
Please keep linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -144,64 +134,23 @@ public JabRefFrame(Stage mainStage,
+
+        this.pushToApplicationCommand = new GuiPushToApplicationCommand(stateManager, dialogService, this.preferences, taskExecutor);
+
+        this.fileHistory = new FileHistoryMenu(this.preferences.getLastFilesOpenedPreferences().getFileHistory(), dialogService, getOpenDatabaseAction());
```
Please keep linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -220,38 +169,9 @@ public JabRefFrame(Stage mainStage,
-                gitHandlerRegistry);
+        MainToolBar mainToolBar = new MainToolBar(this, pushToApplicationCommand, globalSearchBar, dialogService, stateManager, preferences, aiService, fileUpdateMonitor, taskExecutor, entryTypesManager, clipBoardManager, undoManager);
+
+        MainMenu mainMenu = new MainMenu(this, fileHistory, sidePane, pushToApplicationCommand, preferences, stateManager, fileUpdateMonitor, taskExecutor, dialogService, Injector.instantiateModelOrService(JournalAbbreviationRepository.class), entryTypesManager, undoManager, clipBoardManager, this::getOpenDatabaseAction, aiService, entryEditor, gitHandlerRegistry);
```
Please keep linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -402,8 +318,7 @@ private void initBindings() {
         // Wait for the scene to be created, otherwise focusOwnerProperty is not provided
-        Platform.runLater(() -> stateManager.focusOwnerProperty().bind(
-                EasyBind.map(mainStage.getScene().focusOwnerProperty(), Optional::ofNullable)));
+        Platform.runLater(() -> stateManager.focusOwnerProperty().bind(EasyBind.map(mainStage.getScene().focusOwnerProperty(), Optional::ofNullable)));
```
Please keep linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -427,16 +342,13 @@ private void initBindings() {
                 // Set window title dynamically
-                mainStage.titleProperty().bind(Bindings.createStringBinding(
-                        () -> libraryTab.textProperty().getValue() + " – " + FRAME_TITLE, // not a minus, but codepoint 2013
+                mainStage.titleProperty().bind(Bindings.createStringBinding(() -> libraryTab.textProperty().getValue() + " – " + FRAME_TITLE, // not a minus, but codepoint 2013
```
Please keep linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -427,16 +342,13 @@ private void initBindings() {
-                    boolean wasClosed = tabbedPane.getTabs().stream()
-                                                  .filter(tab -> tab instanceof LibraryTab)
-                                                  .noneMatch(ltab -> ((LibraryTab) ltab).getBibDatabaseContext().getUid().equals(activeUID));
+                    boolean wasClosed = tabbedPane.getTabs().stream().filter(tab -> tab instanceof LibraryTab).noneMatch(ltab -> ((LibraryTab) ltab).getBibDatabaseContext().getUid().equals(activeUID));
```
Please keep linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -465,21 +375,9 @@ private void initBindings() {
-                                .orElse(new SimpleBooleanProperty(false))
-                )
-        );
+        stateManager.canGoBackProperty().bind(stateManager.activeTabProperty().flatMap(optionalTab -> optionalTab.map(LibraryTab::canGoBackProperty).orElse(new SimpleBooleanProperty(false))));
```
Please keep linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -465,21 +375,9 @@ private void initBindings() {
-        );
+        stateManager.canGoBackProperty().bind(stateManager.activeTabProperty().flatMap(optionalTab -> optionalTab.map(LibraryTab::canGoBackProperty).orElse(new SimpleBooleanProperty(false))));
+
+        stateManager.canGoForwardProperty().bind(stateManager.activeTabProperty().flatMap(optionalTab -> optionalTab.map(LibraryTab::canGoForwardProperty).orElse(new SimpleBooleanProperty(false))));
```
Please keep linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -510,8 +408,7 @@ private void updateTabBarVisible() {
     public LibraryTab getCurrentLibraryTab() {
-        if (tabbedPane.getSelectionModel().getSelectedItem() == null
-                || !(tabbedPane.getSelectionModel().getSelectedItem() instanceof LibraryTab)) {
+        if (tabbedPane.getSelectionModel().getSelectedItem() == null || !(tabbedPane.getSelectionModel().getSelectedItem() instanceof LibraryTab)) {
```
Please keep linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -557,18 +439,7 @@ public void showWelcomeTab() {
-                undoManager,
-                clipBoardManager,
-                taskExecutor);
+        LibraryTab libraryTab = LibraryTab.createLibraryTab(databaseContext, this, dialogService, aiService, preferences, stateManager, fileUpdateMonitor, entryTypesManager, undoManager, clipBoardManager, taskExecutor);
```
Please keep linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -587,17 +458,10 @@ private ContextMenu createTabContextMenuFor(LibraryTab tab) {
-                factory.createMenuItem(StandardActions.CLOSE_LIBRARY, new CloseDatabaseAction(this, tab, stateManager)),
-                factory.createMenuItem(StandardActions.CLOSE_OTHER_LIBRARIES, new CloseOthersDatabaseAction(tab)),
-                factory.createMenuItem(StandardActions.CLOSE_ALL_LIBRARIES, new CloseAllDatabaseAction()));
+        contextMenu.getItems().addAll(factory.createMenuItem(StandardActions.LIBRARY_PROPERTIES, new LibraryPropertiesAction(tab::getBibDatabaseContext, stateManager)), factory.createMenuItem(StandardActions.OPEN_DATABASE_FOLDER, new OpenDatabaseFolder(dialogService, stateManager, preferences, tab::getBibDatabaseContext)), factory.createMenuItem(StandardActions.OPEN_CONSOLE, new OpenConsoleAction ...
```
Please keep linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -612,10 +476,7 @@ public boolean closeTab(LibraryTab tab) {
-                                       .distinct()
-                                       .filter(getLibraryTabs()::contains)
-                                       .toList();
+        List<LibraryTab> toClose = tabs.stream().distinct().filter(getLibraryTabs()::contains).toList();
```
Please keep linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -753,22 +604,20 @@ public CloseOthersDatabaseAction(LibraryTab libraryTab) {
-                if (libraryTab != toKeepLibraryTab) {
-                    Platform.runLater(() -> closeTab(libraryTab));
-                }
+            List<LibraryTab> libraryTabs = tabbedPane.getTabs().stream().filter(LibraryTab.class::isInstance).map(LibraryTab.class::cast).filter(tab -> tab != toKeepLibraryTab).toList();
```
Please introduce linebreaks

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/MainMenu.java:
```diff
@@ -173,6 +176,35 @@ private void createMenu() {
+                            frame.new CloseOthersDatabaseAction(currentTab).execute();
+                        }
+                    }
+                }),
```
Please extract

calixtus reviewed (dismissed):
Many unneccessary linebreak changes, please fix before we make a review.

jabref-machine commented:
Your code currently does not meet [JabRef's code guidelines](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). We use [Checkstyle](https://checkstyle.sourceforge.io/) to identify issues. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Source Code Tests / Checkstyle (pull_request)" and click on it.

In case of issues with the import order, double check that you [activated Auto Import](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#enable-proper-import-cleanup). You can trigger fixing imports by pressing <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> to trigger [Optimize Imports](https://www.jetbrains.com/guide/tips/optimize-imports/).

Please carefully follow [the setup guide for the codestyle](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). Afterwards, please [run checkstyle locally](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#run-checkstyle) and fix the issues, commit, and push.


<!-- ghprcomment
Source Code Tests
Checkstyle
-->

calixtus commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -118,17 +118,8 @@ private enum PanelMode { MAIN_TABLE, MAIN_TABLE_AND_ENTRY_EDITOR }
-                       ClipBoardManager clipBoardManager,
-                       TaskExecutor taskExecutor,
-                       GitHandlerRegistry gitHandlerRegistry) {
+    @SuppressWarnings("checkstyle:EmptyLineSeparator")
```
Nope.

SatyamKumarCS commented:
Hi, @calixtus! Thanks for the review.
Previously, I had added the line breaks as requested, but I’m having trouble identifying which extra or unnecessary line breaks I introduced. Could you please point me to the exact part of the file or the specific lines where the issue occurs? This will help me take corrective measures.

Thanks again for your guidance!

SatyamKumarCS commented:
@calixtus @koppor Please check now, I have made all the changes requested.

koppor commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -144,8 +145,7 @@ public JabRefFrame(Stage mainStage,
         // Create components
-        this.viewModel = new JabRefFrameViewModel(
-                preferences,
+        this.viewModel = new JabRefFrameViewModel(preferences,
```
Why the reformat?

koppor commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -128,7 +128,8 @@ public JabRefFrame(Stage mainStage,
                        TaskExecutor taskExecutor,
-                       GitHandlerRegistry gitHandlerRegistry) {
+                       GitHandlerRegistry gitHandlerRegistry
+    ) {
```
Why the reformat?

koppor commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -155,22 +155,23 @@ public JabRefFrame(Stage mainStage,
                 this::getOpenDatabaseAction,
-                stateManager);
+                stateManager
+        );
```
Why the reformat?

koppor commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -155,22 +155,23 @@ public JabRefFrame(Stage mainStage,
 
-        this.globalSearchBar = new GlobalSearchBar(
-                this,
+        this.globalSearchBar = new GlobalSearchBar(this,
```
Why the reformat?

koppor commented on jabgui/src/main/java/org/jabref/gui/frame/JabRefFrame.java:
```diff
@@ -155,22 +155,23 @@ public JabRefFrame(Stage mainStage,
                 dialogService,
-                SearchType.NORMAL_SEARCH);
+                SearchType.NORMAL_SEARCH
+        );
```
Why the reformat?

koppor commented:
> @calixtus @koppor Please check now, I have made all the changes requested.

No. Please respect our wishes that we want to have semantic changes only; not just syntactic changes.

jabref-machine commented:
You ticked that you modified `CHANGELOG.md`, but no new entry was found there.

If you made changes that are visible to the user, please add a brief description along with the issue number to the `CHANGELOG.md` file. If you did not, please replace the cross (`[x]`) by a slash (`[/]`) to indicate that no `CHANGELOG.md` entry is necessary. More details can be found in our [Developer Documentation about the changelog](https://devdocs.jabref.org/decisions/0007-human-readable-changelog.html).


<!-- ghprcomment
Check PR CHANGELOG.md
CHANGELOG.md needs to be modified if indicated
-->
